app/javascript/packs/articlePage.jsx

```diff
@@ -12,56 +12,97 @@ if (fullscreenActionElements) {
 
 // The Snackbar for the article page
 const snackZone = document.getElementById('snack-zone');
-
 if (snackZone) {
   render(<Snackbar lifespan="3" />, snackZone);
 }
 
 // eslint-disable-next-line no-restricted-globals
 top.addSnackbarItem = addSnackbarItem;
 
+// Pin/Unpin article
+// these element are added by initializeBaseUserData.js:addRelevantButtonsToArticle
+const toggleArticlePin = async (button) => {
+  const isPinButton = button.id === 'js-pin-article';
+  const { articleId, path } = button.dataset;
+  const method = isPinButton ? 'PUT' : 'DELETE';
+  const body = method === 'PUT' ? JSON.stringify({ id: articleId }) : null;
+
+  const response = await fetch(path, {
+    method,
+    body,
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  });
+
+  // response could potentially fail if the article is draft but we don't show
+  // the buttons in those cases, so I think there's no need to handle that scenario client side
+  if (response.ok) {
+    // replace id and label
+    button.id = isPinButton ? 'js-unpin-article' : 'js-pin-article';
+    button.innerHTML = `${isPinButton ? 'Unpin' : 'Pin'} Post`;
+
+    const message = isPinButton
+      ? 'The post has been succesfully pinned'
+      : 'The post has been succesfully unpinned';
+    addSnackbarItem({ message });
+  }
+};
+
+const actionsContainer = document.getElementById('action-space');
+const pinTargets = ['js-pin-article', 'js-unpin-article'];
+actionsContainer.addEventListener('click', async (event) => {
+  if (pinTargets.includes(event.target.id)) {
+    toggleArticlePin(event.target);
+  }
+});
```

Comment:
I've always been a fan of event bubbling, as it has the advantage of having 1 central handler function present in a page, instead of  N anonymous functions per each button. Granted, we could simply declare it as an async function and attach the same handler to all the buttons. I'll see if I can alter it without too much effort!

Reply:
Mmm now that I think about it, we already share `toggleArticlePin` for the buttons, there's not much difference in doing "for each button, attach the anonymous event handler" vs "listen to this event on the container" and since we swap the ID at runtime, I think this should be fine